nuxeo-ai-core/src/main/java/org/nuxeo/ai/auto/AutoServiceImpl.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * (C) Copyright 2019 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Gethin James
+ */
+package org.nuxeo.ai.auto;
+
+import static org.nuxeo.ai.AIConstants.AUTO_CORRECTED;
+import static org.nuxeo.ai.AIConstants.AUTO_FILLED;
+import static org.nuxeo.ai.auto.AutoService.AUTO_ACTION.ALL;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ai.metadata.AIMetadata;
+import org.nuxeo.ai.metadata.SuggestionMetadataAdapter;
+import org.nuxeo.ai.services.DocMetadataService;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.model.Property;
+import org.nuxeo.ecm.core.api.model.PropertyNotFoundException;
+import org.nuxeo.runtime.api.Framework;
+
+/**
+ * Autofill and AutoCorrect services
+ */
+public class AutoServiceImpl implements AutoService {
+
+    private static final Logger log = LogManager.getLogger(AutoServiceImpl.class);
+
+    @Override
+    public void calculateProperties(DocumentModel doc) {
+        calculateProperties(doc, ALL);
+    }
+
+    @Override
+    public void calculateProperties(DocumentModel doc, AUTO_ACTION action) {
+        SuggestionMetadataAdapter adapted = doc.getAdapter(SuggestionMetadataAdapter.class);
+        for (String xpath : adapted.getSuggestedProperties()) {
+            switch (action) {
+                case ALL:
```

Comment:
Aren't they exclusive ?
Auto-fill if no prior data is there, auto-correct if prior data is there.

Reply:
Elsewhere in the code I mention auto-fill and auto-correct are mutually exclusive, but we still need to call them.  In the code, auto-correct is called before auto-fill - so autofill won't be calculated if autocorrect has set a value.

Reply:
We need the possibliity to run both